engine/src/scheme.rs

```diff
@@ -249,11 +301,20 @@ impl<'s> Scheme {
         Ok(scheme)
     }
 
-    pub(crate) fn get_field_index(&'s self, name: &str) -> Result<Field<'s>, UnknownFieldError> {
+    pub(crate) fn get_field(&'s self, name: &str) -> Result<Field<'s>, UnknownFieldError> {
+        self.get_field_with_path(name, std::iter::empty::<FieldPathItem>())
+    }
+
+    pub(crate) fn get_field_with_path(
+        &'s self,
+        name: &str,
+        path: impl IntoIterator<Item = FieldPathItem>,
+    ) -> Result<Field<'s>, UnknownFieldError> {
         match self.fields.get_full(name) {
             Some((index, ..)) => Ok(Field {
                 scheme: self,
                 index,
+                path: path.into_iter().collect::<Vec<_>>(),
```

Comment:
This seems to be doing implicit allocation in a hot path, probably not a great idea. As mentioned above, I'd rather move `["..."]` to unary ops which would change this part too.